src/TextInputWidget.cc

```diff
@@ -135,28 +135,23 @@ FilteredTextEdit::canInsertFromMimeData(const QMimeData *source) const
 void
 FilteredTextEdit::insertFromMimeData(const QMimeData *source)
 {
-        if (source->hasImage()) {
-                const auto formats = source->formats();
-                const auto idx     = formats.indexOf(
-                  QRegularExpression{"image/.+", QRegularExpression::CaseInsensitiveOption});
-
-                // Note: in the future we may want to look into what the best choice is from the
-                // formats list. For now we will default to PNG format.
-                QString type = "png";
-                if (idx != -1) {
-                        type = formats.at(idx).split('/')[1];
-                }
+        const auto formats = source->formats().filter("/");
+        const auto image   = formats.filter("image/", Qt::CaseInsensitive);
+        const auto audio   = formats.filter("audio/", Qt::CaseInsensitive);
+        const auto video   = formats.filter("video/", Qt::CaseInsensitive);
 
-                // Encode raw pixel data of image.
-                QByteArray data = source->data("image/" + type);
-                previewDialog_.setImageAndCreate(data, type);
-                previewDialog_.show();
-        } else if (source->hasFormat("x-special/gnome-copied-files") &&
-                   QImageReader{source->text()}.canRead()) {
+        if (source->hasImage()) {
+                showPreview(source, image, "image", "png");
+        } else if (source->hasFormat("x-special/gnome-copied-files") ||
+                   QFileInfo{source->text()}.exists()) {
+                // Generic file for any platform.
                 // Special case for X11 users. See "Notes for X11 Users" in source.
                 // Source: http://doc.qt.io/qt-5/qclipboard.html
-                previewDialog_.setImageAndCreate(source->text());
-                previewDialog_.show();
+                previewDialog_.setPreview(source->text());
+        } else if (!audio.empty()) {
+                showPreview(source, audio, "audio", "flac");
+        } else if (!video.empty()) {
+                showPreview(source, video, "video", "webm");
```

Comment:
Why bother to pass in a default media type, when the passed-in formats list in every call is both non-empty and contain only appropriate values? It should be safe to e.g. always select the first.

Reply:
In the previous PR, I recall we were discussing which default image format we should fall-back to. I applied the same concept here. If you feel that having a default fall-back is unnecessary and it is always safe to select the first of the specific formats list (the filtered ones), then I will remove the extra parameters.

Reply:
This code as written will *never* use the "defaults" for anything. It's not clear to me what problem they're intended to solve. Dead code is a maintenance hazard.

Reply:
OK I will do without it. Thanks for your attention to detail!